### PR TITLE
add preflight task

### PR DESCRIPTION
**Pull request by skattoju**
This is to add preflight as a task to the RHTAP pipeline. Preflight is a tool that allows partners to check if their workloads are ready for certification. It was raised by a partner that RH content is sometimes not certification ready. Given that that RHTAP will eventually become the standard way to build RH components. It is proposed that preflight should be part of the red hat trusted application pipeline.

It will be a non blocking check. Whether the check is enforced will be determined by a downstream policy or enterprise contract.

Update:
The TEST_OUTPUT issue was resolved. The problem was that the results were too big to fit in tekton task output vars. Currently the user can checkout the output in the logs and only see the result in the output variable...

### CODE REVIEW

**Comment by jsztuka** [comment]
Hello, I was wondering if this change has story connected to it that could be referenced(supposedly Jira story) to know more background, like team originator, reason behind this check etc.

**Comment by jsztuka** [comment]
I will refer here for a slack channel discussion for better overview of this PR:
https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1693507130582899

**Comment by skattoju** [comment]
@jsztuka I have created https://issues.redhat.com/browse/RHTAP-1591 lmk if additional information is needed

File: task/preflight/0.1/preflight.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    - name: image-url
+      description: Image url to scan.
+  results:
+    - name: TEST_OUTPUT
+      description: Preflight pass or fail outcome.
+    - name: PFLT_OUTPUT
+      description: Preflight check summary.
+  steps:
+    - name: check-container
+      image: quay.io/opdev/preflight:stable
```
Comment (MartinBasti) [inline task/preflight/0.1/preflight.yaml:18]: please pin image to digest
quay.io/opdev/preflight:stable@sha256:9a868eb7f2d6f4c5a14e0da03f9a2342ee726c62ba2f129d0ca29a302c5d67d6

File: task/preflight/0.1/preflight.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      volumeMounts:
+        - name: pfltoutputdir
+          mountPath: /artifacts
+    - name: gather-pflt-results
+      image: quay.io/redhat-appstudio/appstudio-utils:cb1a8a07fa536d0f661990ff06e10954895d448a
+      volumeMounts:
+        - name: pfltoutputdir
+          mountPath: /artifacts
+      script: |
+        #expected results directory
```
Comment (MartinBasti) [inline task/preflight/0.1/preflight.yaml:29]:
```suggestion
        # expected results directory
```

File: task/preflight/0.1/preflight.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        RESULT_JSON_PATH=artifacts/results.json
+
+        #Process results
```
Comment (MartinBasti) [inline task/preflight/0.1/preflight.yaml:32]:
```suggestion
        # Process results
```

File: task/preflight/0.1/preflight.yaml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        cat $RESULT_JSON_PATH | tee $(results.PFLT_OUTPUT.path)
+        PFLT_RESULT="FAILURE"
+        PFLT_PASSED=$(cat $RESULT_JSON_PATH |jq .passed)
+        if [ $PFLT_PASSED = true ]; then PFLT_RESULT="SUCCESS" ; fi
+        PFLT_PASS_COUNT=$(cat $RESULT_JSON_PATH | jq '.results.passed | length')
+        PFLT_FAIL_COUNT=$(cat $RESULT_JSON_PATH | jq '.results.failed |length')
+        PFLT_ERROR_COUNT=$(cat $RESULT_JSON_PATH | jq '.results.errors |length')
+        if [[ $PFLT_ERROR_COUNT > 0 ]]; then PFLT_RESULT="ERROR" ; fi
+
+        #Generate TEST_OUTPUT
```
Comment (MartinBasti) [inline task/preflight/0.1/preflight.yaml:42]:
```suggestion
        # Generate TEST_OUTPUT
```

File: task/preflight/0.1/preflight.yaml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        TEST_OUTPUT=$(jq -rce \
```
Comment (MartinBasti) [inline task/preflight/0.1/preflight.yaml:43]: Output should have also  note to inform users with short summary

Comment (skattoju) [reply task/preflight/0.1/preflight.yaml:44]: I thought it was optional.. but i have added it 👍

Comment (skattoju) [reply task/preflight/0.1/preflight.yaml:44]: @MartinBasti I still get green 🟢 even when the result says FAILURE. I want to be 🟡 when its not SUCCESS. The code appears to be there but may be its not deployed in preview. Is there any other environment i can test in ?

Comment (MartinBasti) [reply task/preflight/0.1/preflight.yaml:44]: Green where? in openshift pipelines or in RHTAP web interface?
I saw it working in RHTAP UI

File: task/preflight/0.1/preflight.yaml (hunk 1 above)
Comment (MartinBasti) [reply task/preflight/0.1/preflight.yaml]: you omitted `stable` tag, please keep it there together with digest

Comment (skattoju) [reply task/preflight/0.1/preflight.yaml]: oops. added. thanks for the feedback 👍

File: task/preflight/0.1/preflight.yaml (hunk 5 above)
Comment (skattoju) [reply task/preflight/0.1/preflight.yaml:44]: In the UI of RHTAP preview but my guess is that it does not have the latest code ..

**Comment by MartinBasti** [comment]
LGTM, but I'm still waiting for confirmation if this can be part of default pipelines

**Comment by ralphbean** [comment]
@MartinBasti yes, we caught this during JIRA weeding today (found via https://issues.redhat.com/browse/RHTAP-1591).

From talking with PM, this is something we want as a part of default build pipelines.

**Review by MartinBasti — APPROVED** [review]
Please rebase

**Review by dirgim — APPROVED** [review]
LGTM

**Review by hongweiliu17 — APPROVED** [review]
LGTM

**Comment by skattoju** [comment]
@MartinBasti there was a discussion around renaming preflight to ecosystem-cert-preflight-checks or something like that thats no longer needed ?

**Comment by MartinBasti** [comment]
Sorry I thought that name was already decided

File: task/preflight/0.1/preflight.yaml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      description: Preflight check summary.
+  steps:
+    - name: check-container
+      image: quay.io/opdev/preflight:stable@sha256:d990cbe8fb0db4dff0f7609403992f273f07d55f89dc7d2f046eb8f9c3e48110
+      args: ["check", "container", "$(params.image-url)"]
+      volumeMounts:
+        - name: pfltoutputdir
+          mountPath: /artifacts
+    - name: gather-pflt-results
+      image: quay.io/redhat-appstudio/appstudio-utils:cb1a8a07fa536d0f661990ff06e10954895d448a
```
Comment (MartinBasti) [inline task/preflight/0.1/preflight.yaml:24]: this image should have also digest, I just noticed it looked like a digest from distance but it's just a tag

Comment (skattoju) [reply task/preflight/0.1/preflight.yaml]: Sorry for the delay. I have updated it to a digest.

File: pipelines/template-build/template-build.yaml [hunk 7]
```diff
@@ -162,6 +162,19 @@ spec:
...
         value: $(tasks.build-container.results.IMAGE_URL)
+    - name: preflight
+      when:
+      - input: $(params.skip-checks)
+        operator: in
+        values: ["false"]
+      runAfter:
+        - build-container
+      taskRef:
+        name: preflight
```
Comment (tisutisu) [inline pipelines/template-build/template-build.yaml:173]: Do we have to reference the task using name `ecosystem-cert-preflight-checks` here?

Comment (skattoju) [reply pipelines/template-build/template-build.yaml:173]: Thanks for pointing it out.. Totally missed that!

**Pull request merged**